File: day4/day4.rb
# HEEEY Leaderboard 72
open('day4.txt')
	.each_line
	.map(&:chomp)
	.chunk_while { !_2.empty? }
	.map { _1.reject(&:empty?) }
	.map { _1.join(' ').split(/[: ]/).each_slice(2).to_h }
	.each { _1.delete 'cid' }
	.tap { |stream|
		stream.map(&:keys)
			.map(&:sort)
			.each { puts _1 - %w(byr ecl eyr hcl hgt iyr pid) }
			.map { %w(byr ecl eyr hcl hgt iyr pid) - _1 }
			.count(&:empty?)
			.tap { puts "Part 1: #{_1}" }
	}
	.select { ('1920'..'2002').include? _1['byr'] }
	.select { ('2010'..'2020').include? _1['iyr'] }
	.select { ('2020'..'2030').include? _1['eyr'] }
	.select { _1['hgt'] =~ /(\d+)(cm|in)/ and ($2 == 'cm' ? 150..193 : 59..76).include? $1.to_i }
	.select { _1['hcl'] =~ /#\h{6}/ }
	.select { %w(amb blu brn gry grn hzl oth).include? _1['ecl'] }
	.select { _1['pid'] =~ /\d{9}/ }
	.count
	.tap { puts "Part 2: #{_1}" }

__END__
# Here's the un-cleaned-up code I used 
open('day4.txt')
	.each_line
	.map(&:chomp)
	.chunk_while { !_2.empty? }
	.map { _1.reject(&:empty?) }
	.map { _1.join(' ').split(/[: ]/).each_slice(2).to_h }
	.each { _1.delete 'cid' }
	.select { (1920..2002).include? _1['byr'].to_i }
	.select { (2010..2020).include? _1['iyr'].to_i }
	.select { (2020..2030).include? _1['eyr'].to_i }
	.select { _1['hgt'] =~ /^(\d+)(cm|in)$/ and ($2 == 'cm' ? 150..193 : 59..76).include? $1.to_i }
	.select { _1['hcl'] =~ /^#\h{6}$/ }
	.select { %w(amb blu brn gry grn hzl oth).include? _1['ecl'] }
	.select { _1['pid'] =~ /^\d{9}$/ }
	.count
	.display

File: 2020/day4/day4.py
import sys, itertools, re

file = 'day4.txt' if len(sys.argv) < 2 else sys.argv[1]
part1 = 0
part2 = 0

with open(file) as f:
	acc = {}
	for line in itertools.chain(map(str.strip, f), ['']):
		if line:
			acc.update(ele.split(':') for ele in line.split(' '))
			continue

		passport, acc = acc, {}

		if {'pid', 'hgt', 'ecl', 'byr', 'eyr', 'iyr', 'hcl'} - set(passport):
			continue

		part1 += 1

		if int(passport['byr']) not in range(1920, 2004):
			continue

		if int(passport['iyr']) not in range(2010, 2021):
			continue

		if int(passport['eyr']) not in range(2020, 2031):
			continue

		if passport['ecl'] not in {'amb', 'blu', 'brn', 'gry', 'grn', 'hzl', 'oth'}:
			continue

		if not re.match(r'\d{9}', passport['pid']):
			continue

		if not re.match(r'#[\da-fA-F]', passport['hcl']):
			continue

		m = re.match(r'(\d+)(cm|in)', passport['hgt'])
		if m and int(m[1]) in (range(150, 194) if m[2] == 'cm' else range(59, 77)):
			part2 += 1

print('Part 1:', part1)
print('Part 2:', part2)
